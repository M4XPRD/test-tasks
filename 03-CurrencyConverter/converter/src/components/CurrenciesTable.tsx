import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import { Country } from '../slices/countriesSlice';
import { RootState } from '../slices';

const formRow = (country: object, curRates?: Record<string, number> | undefined, type?: string) => {
  const countryType = country as Country;
  const [currency] = Object.keys(countryType.currencies);
  const currencyName = countryType.currencies[currency].name;
  switch (type) {
    case 'currency':
      return `${currency} — ${currencyName}`;
    case 'rates':
      return curRates && curRates[currency] ? curRates[currency].toFixed(2) : '—';
    default:
      return countryType.name.common;
  }
};

const CurrenciesTable = () => {
  const countriesList = useSelector((state: RootState) => state.countries.countriesList);
  const exchangeRates = useSelector((state: RootState) => state.currencies.exchange.rates.rates);
  const { t } = useTranslation();

  return (
    <TableContainer className="currencies-table-container">
      <Table className="currencies-table">
        <TableHead>
          <TableRow className="table-head-row">
            <TableCell align="center" width="33%">{t('components.currenciesTable.country')}</TableCell>
            <TableCell align="center" width="33%">{t('components.currenciesTable.currency')}</TableCell>
            <TableCell align="center" width="33%">{t('components.currenciesTable.rate')}</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {countriesList.map((country) => (
            <TableRow
              className="table-body-row"
              key={country.name.common}
            >
              <TableCell align="left" component="th" scope="row">
                {formRow(country)}
              </TableCell>
              <TableCell align="left" component="th" scope="row">{formRow(country, exchangeRates, 'currency')}</TableCell>
              <TableCell align="center" component="th" scope="row">{formRow(country, exchangeRates, 'rates')}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default CurrenciesTable;
