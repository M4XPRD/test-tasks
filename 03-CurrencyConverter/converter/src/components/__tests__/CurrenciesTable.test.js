import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../slices/index';

import CurrenciesTable from '../CurrenciesTable';

describe('CurrenciesTable component', () => {
  it('CurrenciesTable snapshot', () => {
    const element = render(
      <Provider store={store}>
        <CurrenciesTable />
      </Provider>,
    );

    expect(element).toMatchSnapshot();
  });
});
