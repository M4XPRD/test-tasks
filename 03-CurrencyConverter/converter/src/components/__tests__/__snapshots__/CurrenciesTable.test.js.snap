// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrenciesTable component CurrenciesTable snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiTableContainer-root currencies-table-container css-rorn0c-MuiTableContainer-root"
      >
        <table
          class="MuiTable-root currencies-table css-rqglhn-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head table-head-row css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                scope="col"
                width="33%"
              >
                components.currenciesTable.country
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                scope="col"
                width="33%"
              >
                components.currenciesTable.currency
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                scope="col"
                width="33%"
              >
                components.currenciesTable.rate
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiTableContainer-root currencies-table-container css-rorn0c-MuiTableContainer-root"
    >
      <table
        class="MuiTable-root currencies-table css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head table-head-row css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
              width="33%"
            >
              components.currenciesTable.country
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
              width="33%"
            >
              components.currenciesTable.currency
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
              width="33%"
            >
              components.currenciesTable.rate
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
